### ISSUE TITLE
Catch exception into an Either

### ISSUE THREAD

**Issue by jesseschalken** [arrow-kt/arrow#734]
I expected to find a function that will run a function, or a `Try`, and catch a specified type of exception as the left side of an `Either`, but I can't find it anywhere.

Something like this (convert a `Try` to an `Either` for a specific exception type):

```kotlin
inline fun <T, reified E: Throwable> Try<T>.catchEither(): Either<E, T> =
    fold({ if (it is E) Left(it) else throw it }, { Right(it) })
```
```kotlin
fun myParse(json: String): Either<JsonParseException, MyObject> = 
    Try { Gson().fromJson<MyObject>(json) }.catchEither()
```

or this (run a function catching a specific exception type into an `Either`):

```kotlin
inline fun <T, reified E: Throwable> TryCatch(block: () -> T): Either<E, T> =
    try { Right(block()) } catch (e: Throwable) { if (e is E) Left(e) else throw e }
```
```kotlin
fun myParse(json: String): Either<JsonParseException, MyObject> = 
    TryCatch { Gson().fromJson<MyObject>(json) }
```

Is there already a way to do this? If not, would the authors consider adding it?

Thanks

**Comment by jesseschalken** [comment]
I also can't seem to find a way to go the other way. Eg:

```kotlin
fun <T> Either<Throwable, T>.orThrow(): T = fold({ throw it }, { it })
```
```kotlin
val object = myParse("{ some json }").orThrow()
```

**Comment by pakoito** [comment]
`Try { myParse("{ some json }") }.toEither()`

https://github.com/arrow-kt/arrow/blob/master/modules/core/arrow-core/src/main/kotlin/arrow/core/Try.kt#L134

Is that what you want?

**Comment by pakoito** [comment]
>and catch a specified type of exception as the left side of an Either

This is the tricky part. We do not provide operators that throw internally, as they're unsafe and not in the style of FP so your implementation of `catchEither` could not be part of the lib.

What I suggest instead is to `fold` the `Try` into a sensible value for unwanted exceptions. Something like:

```
sealed class DomainExceptions<E>

data class Known<E> (val e: E): DomainExceptions<E>

data class Unknown(val t: Throwable): DomainExceptions<Nothing>
```

Try { myParse("{ some json }") }.fold({ if (it is E) { Either.Left(Known(it)) } else { Either.Left(Unknown(it)) } }, { Either.Right(it) })

**Comment by jesseschalken** [comment]
> We do not provide operators that throw internally

Yes you do. [`Try.get()` throws if it is a `Try.Failure`](https://github.com/arrow-kt/arrow/blob/5da58615131b2f0eceea3cb66cbe9a9821dee461/modules/core/arrow-core/src/main/kotlin/arrow/core/Try.kt#L148), for example. Or is that allowed because it's just re-throwing an exception that was thrown by code outside of the lib? If so then the same applies to my `tryEither` and `TryCatch` functions.

> as they're unsafe and not in the style of FP

I agree. They're unsafe in the sense that they're not statically checked since Kotlin removed Java's checked exceptions. That's why I'm trying to use `Either<E, T>` where possible, but I need a concise way of translating between the `Either` monadic FP world and the exception-throwing Kotlin/Java world. `Try { ... }.toEither()` provides half the story, but produces an `Either<Throwable, T>` so if I know that the function is only going to throw a specific subtype of `Exception`, that information is lost.

**Comment by pakoito** [comment]
>Yes you do. Try.get() throws if it is a Try.Failure, for example.
It's marked as deprecated. It's a left-over from when we merged with FunKtionale. It's going away soon.
https://github.com/arrow-kt/arrow/blob/5da58615131b2f0eceea3cb66cbe9a9821dee461/modules/core/arrow-core/src/main/kotlin/arrow/core/Try.kt#L114-L115

**Comment by pakoito** [comment]
@jesseschalken have you looked into the other solution I proposed? What it does is convert your errors from known types, while also covering unknown types. That way the function is safe and you get what you need :D

https://github.com/arrow-kt/arrow/issues/734#issuecomment-372968642

**Comment by jesseschalken** [comment]
> It's marked as deprecated.

Ah, it is too. Sounds like you don't want Arrow to be throwing any exceptions, even exceptions that were originally thrown from outside of the library.

> have you looked into the other solution I proposed?

I'm happy using the functions I posted above, I was just hoping they already existed in Arrow in some form, or could be added to help users translate between `Either` and throwing exceptions, but if that's not within scope of the library that's fine.

**Comment by pakoito** [comment]
The gist of it is moving away from modeling errors as exceptions. Exceptions have an indeterminate number of possible inheritors, most of them outside your control. Regular try/catch does the job well for those. If you're hoping to move into FP you limit exceptions to the ones gotten from external APIs, and treat them as I did on my example by partitioning their space between the ones you want and the ones you don't. With that you can now control how many possible `Exceptions()` you can control, and which ones you don't, as it's expected from FP with sealed classes.

**Comment by pakoito** [comment]
@jesseschalken do you believe this issue was solved? I would like to close it :)

**Comment by maxcruz** [comment]
I have this function to do something similar when working with suspended functions. 
```
suspend inline fun <reified L : Exception, R> Either.Companion.catchTyped(crossinline block: suspend () -> R): Either<L, R> =
    catch { block() }
        .mapLeft {
            when(it) {
                is L -> it
                else -> throw it
            }
        }
```
